include/rmm/mr/device/arena_memory_resource.hpp

```diff
@@ -0,0 +1,287 @@
+ * 1-116). Springer, Berlin, Heidelberg.
+ * \see Berger, E. D., McKinley, K. S., Blumofe, R. D., & Wilson, P. R. (2000). Hoard: A scalable
+ * memory allocator for multithreaded applications. ACM Sigplan Notices, 35(11), 117-128.
+ * \see Evans, J. (2006, April). A scalable concurrent malloc (3) implementation for FreeBSD. In
+ * Proc. of the bsdcan conference, ottawa, canada.
+ * \see https://sourceware.org/glibc/wiki/MallocInternals
+ * \see http://hoard.org/
+ * \see http://jemalloc.net/
+ * \see https://github.com/google/tcmalloc
+ *
+ * @tparam UpstreamResource memory_resource to use for allocating the arenas. Implements
+ * rmm::mr::device_memory_resource interface.
+ */
+template <typename Upstream>
+class arena_memory_resource final : public device_memory_resource {
+ public:
+  /// The required alignment of this allocator.
+  static constexpr std::size_t allocation_alignment = 256;
+
+  /**
+   * @brief Construct an `arena_memory_resource`.
+   *
+   * @throws rmm::logic_error if `upstream_mr == nullptr`
+   *
+   * @param upstream_mr The memory_resource from which to allocate memory for the arenas.
+   */
+  explicit arena_memory_resource(Upstream* upstream_mr) : upstream_mr_{upstream_mr}
+  {
+    RMM_EXPECTS(nullptr != upstream_mr, "Unexpected null upstream pointer.");
+  }
+
+  // Disable copy (and move) semantics.
+  arena_memory_resource(arena_memory_resource const&) = delete;
+  arena_memory_resource& operator=(arena_memory_resource const&) = delete;
+
+  /**
+   * @brief Queries whether the resource supports use of non-null CUDA streams for
+   * allocation/deallocation.
+   *
+   * @returns bool true.
+   */
+  bool supports_streams() const noexcept override { return true; }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool false
+   */
+  bool supports_get_mem_info() const noexcept override { return false; }
+
+ private:
+  using arena      = detail::arena::arena<Upstream>;
+  using read_lock  = std::shared_lock<std::shared_timed_mutex>;
+  using write_lock = std::lock_guard<std::shared_timed_mutex>;
+
+  /**
+   * @brief Allocates memory of size at least `bytes`.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `std::bad_alloc` if the requested allocation could not be fulfilled
+   *
+   * @param bytes The size in bytes of the allocation
+   * @param stream The stream to associate this allocation with
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    if (bytes <= 0) return nullptr;
+
+    bytes = rmm::detail::align_up(bytes, allocation_alignment);
+    RMM_EXPECTS(
+      bytes <= get_maximum_allocation_size(), rmm::bad_alloc, "Maximum allocation size exceeded");
+
+    if (arena::handles_size(bytes)) {
+      return get_arena(stream).allocate(bytes);
+    } else {
+      RMM_ASSERT_CUDA_SUCCESS(cudaStreamSynchronize(stream));
```

Comment:
No longer using the stream-ordered pool. The synchronization is necessary since the global arena is a shared pool for allocating large chunks of memory. It's used by all threads/streams. This is mainly to avoid fragmentation when running under high concurrency.

I think the key observation here is that in a real-world application, the allocation sizes follow some kind of power law distribution, with smaller allocations more common. By handling small allocations in the per-thread arena, we can afford the synchronization cost for allocating large sizes in the global arena. Since it's a single shared arena, we can avoid additional memory fragmentation when running under high concurrency.

Reply:
It sure would be nice to see a demonstration of this rather than an assumption.

Reply:
See benchmarks.

Reply:
The benchmarks are synthetic -- they don't do any work on the GPU, only allocation. So you will not see the impact of stream synchronization on these benchmarks. But I'm guessing that you will see the impact on real world nsys timelines. 

I realize that your per-thread arenas avoid frequent synchronization for very small allocations, which is good. But the arena_MR synchronizes in multiple locations. My intuition is that this would cause a lot of GPU pipeline bubbles. E.g. if a stream allocates a 64MiB or larger array A, copies to it, runs a kernel on it, frees it, and then allocates another array B and runs a kernel on B, there will be a pipeline flush after each allocation, whereas if the allocations were stream ordered, there would be no gaps.